### PR TITLE
feat: merge upstream template v2.3.1

### PR DESCRIPTION
**Pull request by jens-getpassport**
This change updates the package template repo to the [latest update of the public upstream package](https://github.com/jenstroeger/python-package-template/releases). Notable features:

- Replaced the [Semantic Release Action](https://github.com/relekang/python-semantic-release) with built-in use of the [commitizen](https://commitizen-tools.github.io/commitizen/) tool (which we already use in our git hooks);
- [SLSA Level 3](https://slsa.dev/spec/v0.1/levels) compliance, including the generation of a Software Bill of Materials and reproducable build assets attached to a release;
- Refactor of Action workflows to minimize number of runs.
- Heaps of fixes, see the [upstream CHANGELOG](https://github.com/jenstroeger/python-package-template/blob/main/CHANGELOG.md).

Merging this PR should close all existing Dependabot PRs.

Supersedes https://github.com/getpassport/template_package_python/pull/21

### CODE REVIEW

File: .github/dependabot.yaml [hunk 1]
```diff
@@ -14,6 +14,10 @@ updates:
     include: scope
   open-pull-requests-limit: 13
   target-branch: develop
+  # https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#reviewers
```
Comment (brad-getpassport) [inline .github/dependabot.yaml:17]: Maybe add a comment that you can add reviewers with permission?

File: .github/disabled-workflows/codeql-analysis.yaml [hunk 2]
```diff
@@ -33,42 +33,42 @@ jobs:
         language: [python]
         # CodeQL supports [ 'cpp', 'csharp', 'go', 'java', 'javascript', 'python', 'ruby' ]
         # Learn more about CodeQL language support at https://git.io/codeql-language-support
-        python: ['3.9', '3.10']
+        python: ['3.10']
```
Comment (brad-getpassport) [inline .github/disabled-workflows/codeql-analysis.yaml:36]: Will this miss issues with 3.9 or is it still valid for 3.9 issues?

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -21,26 +40,82 @@ jobs:
...
+      # We don't need to check and test the package separately because `make dist` runs
+      # those targets first and only builds the package if they succeed.
+      # Build the sdist and wheel distribution of the package and docs as a zip file.
+      run: make dist
+    - name: Compute package hash
+      if: matrix.os == env.ARTIFACT_OS && matrix.python == env.ARTIFACT_PYTHON
+      id: compute-hash
+      shell: bash
+      run: |
+        set -euo pipefail
```
Comment (brad-getpassport) [inline .github/workflows/build.yaml:92]: Do we want short configs, versus _**long name**_?

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    - name: Check out repository
+      uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b # v3.0.2
+      with:
+        fetch-depth: 0
+        token: ${{ secrets.REPO_ACCESS_TOKEN }}
+    - name: Set up Python
+      uses: actions/setup-python@b55428b1882923874294fa556849718a1d7f2ca5 # v4.2.0
+      with:
+        python-version: '3.10'
+    - name: Set up Commitizen
```
Comment (brad-getpassport) [inline .github/workflows/release.yaml:48]: Maybe add a comment to see the check-and-release.yaml for comments as to why we're doing this -- setting up commitizen separately

File: docs/source/conf.py [hunk 5]
```diff
@@ -19,7 +19,7 @@
 # -- Project information -----------------------------------------------------
 
 project = "Package"
-copyright = "2021, Package"
+copyright = "2022, Package"
```
Comment (brad-getpassport) [inline docs/source/conf.py:22]: I think this should be 2021-2022, we always start with the original date of creation and keep the range, unless you are saying this version isn't copyrighted before 2022...  which I think is wrong...

**Review by brad-getpassport — APPROVED** [review]
Minor comments that I think we should address, but you can do that secondary.

File: .github/workflows/build.yaml (hunk 3 above)
Comment (jens-getpassport) [reply .github/workflows/build.yaml:92]: Unfortunately the [`set`](https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html) built-in doesn’t have long options 😞

File: .github/disabled-workflows/codeql-analysis.yaml (hunk 2 above)
Comment (jens-getpassport) [reply .github/disabled-workflows/codeql-analysis.yaml:36]: We disabled CodeQL analysis but if I remember correctly, it works for all versions of Python. I’ll check with @behnazh.

File: docs/source/conf.py (hunk 5 above)
Comment (jens-getpassport) [reply docs/source/conf.py]: Commit https://github.com/getpassport/template_package_python/pull/30/commits/62461f4ab5d95a0277a2d476bbdbcbf092101f38.

File: .github/workflows/release.yaml (hunk 4 above)
Comment (jens-getpassport) [reply .github/workflows/release.yaml:49]: Commit https://github.com/getpassport/template_package_python/pull/30/commits/59ef35c5748120592dd76e24cc72950df5083ae1.

File: .github/dependabot.yaml (hunk 1 above)
Comment (jens-getpassport) [reply .github/dependabot.yaml:18]: Commit https://github.com/getpassport/template_package_python/pull/30/commits/59ef35c5748120592dd76e24cc72950df5083ae1.

**Review by brad-getpassport — APPROVED** [review]

**Comment by jens-getpassport** [comment]
@brad-getpassport I’m currently fixing up (done, and testing) the Wiki Action to remove the dependency on the now-retired [Wiki Publish](https://github.com/SwiftDocOrg/github-wiki-publish-action) Action.

**Review by brad-getpassport — APPROVED** [review]

**Pull request merged**